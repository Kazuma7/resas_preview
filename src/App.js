import { useEffect, useState } from "react";
import { fetchPrefectures } from "./fetchPrefectures";
import { fetchPrefInfo } from "./fetchPrefInfo";
import { PrefectureCheckbox } from './PrefectureCheckbox';
import { Line } from 'react-chartjs-2';
import prefColor from './prefColor';
import './App.css';
import { Message } from './Message';
import apiKey from './apiKey';

const graphOption = {
  scales: {
    x: {
      display: true,
      title: {
        display: true,
        text: '年代 (年)'
      },
      },
    y: {
      display: true,
      title: {
        display: true,
        text: '総人口 (千人)'
      },
      suggestedMin: 0,
      suggestedMax: 15000000 / 1000,
    },
  },
  //アスペクト比率をデフォルトから変更するため
  responsive: true,
  maintainAspectRatio: false,
};

const App = () => {

  const [prefectures, setPrefectures] = useState({});
  const [prefecturesLoading, setPrefecturesLoading] = useState(false);
  const [selectedPrefectures, setSelectedPrefectures] = useState(Array(47).fill(false));
  const [series, setSeries] = useState([]);
  const [prefInfo, setPrefInfo] = useState([]);

  useEffect(() => {
        const f = async () => {
          setPrefecturesLoading(true)

          const [fetchedPrefName, fetchedPrefInfo] = await Promise.all([
            fetchPrefectures(apiKey),
            fetchPrefInfo(apiKey),
          ]);

          setPrefectures(fetchedPrefName)
          setPrefInfo(fetchedPrefInfo)

          setPrefecturesLoading(false)
        }
        f()
  }, [])

  const changeSelection = (index) => {
    //チェックボックスを連打されたときのバグを防ぐため
    setPrefecturesLoading(true)

    // checkboxのTrue/Falseの配列を複製
    const selected_copy = selectedPrefectures.slice();

    //checkboxが押されたらTrue/Falseを反転
    selected_copy[index] = !selected_copy[index];

    //chechboxにチェックが入ったらグラフ表示用のデータ配列にデータを挿入
    if (!selectedPrefectures[index]) {
      const res_series = {
        data: prefInfo[index],
        label: prefectures[index].prefName,
        //あらかじめ決めておいた色コードを都道府県に合わせて取得(被らない)
        backgroundColor: prefColor[index],
        borderColor: prefColor[index],
      }
      setSelectedPrefectures(selected_copy)
      setSeries([...series, res_series])
      setPrefecturesLoading(false)
      //checkboxが外されたらグラフ表示用のデータ配列からデータを削除
    } else {
      const series_copy = series.slice();
      for (let i = 0; i < series_copy.length; i++) {
        if (series_copy[i].label === prefectures[index].prefName) {
          series_copy.splice(i, 1);
        }
      }

      setSelectedPrefectures(selected_copy)
      setSeries(series_copy)
      setPrefecturesLoading(false)
    }
  }


  return (
    <div>
      <h1>日本総人口比較webサービス</h1>

      <Message />

      <h2>都道府県</h2>

      <div className="checkboxes">
        {Object.keys(prefectures).map(i => {
          // 都道府県のチェックボックスを横に並べて表示させたい
          return (
            <PrefectureCheckbox
              checked={selectedPrefectures[prefectures[i].prefCode - 1]}
              disabled={prefecturesLoading}
              prefName={prefectures[i].prefName}
              prefCode={prefectures[i].prefCode}
              onChange={() => changeSelection(prefectures[i].prefCode - 1)}
            />
          )
        })}
      </div>

      <h2>総人口グラフ</h2>
      <div className="graph">
        <div className="graph-inner">
          <Line
            data={{
              labels: [1960, 1965, 1970, 1975, 1980, 1985, 1990, 1995, 2000, 2005, 2010, 2015, 2020, 2025, 2030, 2035, 2040, 2045],
              datasets: series,
            }}
            options={graphOption}
          />
        </div>
      </div>
    </div>
  )
}

export default App;
