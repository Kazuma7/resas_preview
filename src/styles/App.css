.App {
    text-align: center;
}

.App-logo {
    height: 40vmin;
    pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
    .App-logo {
        animation: App-logo-spin infinite 20s linear;
    }
}

.App-header {
    background-color: #282c34;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: white;
}

.App-link {
    color: #61dafb;
}

@keyframes App-logo-spin {
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
}

* {
    margin: 0px;
    padding: 0px;
    font-family: '"M PLUS 1p"';
}

h1 {
    text-align: center;
    font-size: 1.2rem;
    background-color: #f5f4f1;
    padding: 50px 0%;
    color: #1a1a1a;
}

h2 {
    margin: 5% 4%;
    font-size: 1.2rem;
    border-bottom: 0.5px solid #333333;
}

p {
    font-size: 1rem;
}

a {
    font-size: 1rem;
}

input {
    margin: 0px 2px 1px 0px;
    vertical-align: middle;
}

.checkboxes {
    margin: 0% 4%;
    font-size: 1rem;
}

.main_message-box {
    margin: 4% 4% 1.5% 4%;
    padding: 3%;
    border-radius: 10px;
    font-size: 0.8rem;
    background-color: #f5f4f1;
}

.sub_message_box {
    margin: 14px 4% 28px 4%;
    font-size: 0.8rem;
}

.sub_m_a1 {
    padding-bottom: 8px;
    display: inline-block;
}

.graph {
    width: 80%;
    height: 0;
    position: relative;
    margin: 0% 0% 5% 3%;
    padding-top: 30%;
}

.graph-inner {
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
    left: 0;
}

.footer {
    padding: 50px 0%;
    text-align: center;
    font-size: 0.8rem;
    background-color: #f5f4f1;
    color: #808080;
}

@media (max-width: 425px) {
    .checkboxes {
        font-size: 0.8rem;
    }

    h1 {
        font-size: 1rem;
        padding: 25px 0px;
    }

    h2 {
        font-size: 1rem;
    }

    .footer {
        padding: 30px 0px;
    }

    p {
        font-size: 0.8rem;
    }

    a {
        font-size: 0.8rem;
    }

    .main_message-box {
        margin: 4%;
    }

    .graph {
        width: 90%;
        height: 0;
        position: relative;
        margin: 0% 0% 5% 3%;
        padding-top: 75%;
    }

    .graph-inner {
        width: 100%;
        height: 100%;
        position: absolute;
        top: 0;
        left: 0;
    }
}
