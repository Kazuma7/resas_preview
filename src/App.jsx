import { useEffect, useState } from 'react';
import { fetchPrefectures } from './function/fetchPrefectures';
import { fetchPrefInfo } from './function/fetchPrefInfo';
import { PrefectureCheckbox } from './components/PrefectureCheckbox';
import { Line } from 'react-chartjs-2';
import prefColor from './styles/prefColor';
import './styles/App.css';
import { Message } from './components/Message';
import apiKey from './apiKey';

const graphOption = {
  scales: {
    x: {
      display: true,
      title: {
        display: true,
        text: '年代 (年)',
      },
    },
    y: {
      display: true,
      title: {
        display: true,
        text: '総人口 (千人)',
      },
      suggestedMin: 0,
      suggestedMax: 15000000 / 1000,
    },
  },
  //アスペクト比率をデフォルトから変更するため
  responsive: true,
  maintainAspectRatio: false,
};

const App = () => {
  const [prefectures, setPrefectures] = useState({});
  const [prefecturesLoading, setPrefecturesLoading] = useState(false);
  const [selectedPrefectures, setSelectedPrefectures] = useState(
    Array(47).fill(false)
  );
  const [series, setSeries] = useState([]);
  const [prefInfo, setPrefInfo] = useState([]);

  useEffect(() => {
    const f = async () => {
      setPrefecturesLoading(true);

      const [fetchedPrefName, fetchedPrefInfo] = await Promise.all([
        fetchPrefectures(apiKey),
        fetchPrefInfo(apiKey),
      ]);

      setPrefectures(fetchedPrefName);
      setPrefInfo(fetchedPrefInfo);

      setPrefecturesLoading(false);
    };
    f();
  }, []);

  const changeSelection = (index) => {
    //チェックボックスを連打されたときのバグを防ぐため
    setPrefecturesLoading(true);

    // checkboxのTrue/Falseの配列を複製
    const selectedCopy = selectedPrefectures.slice();

    //checkboxが押されたらTrue/Falseを反転
    selectedCopy[index] = !selectedCopy[index];

    //chechboxにチェックが入ったらグラフ表示用のデータ配列にデータを挿入
    if (!selectedPrefectures[index]) {
      const resSeries = {
        data: prefInfo[index],
        label: prefectures[index].prefName,
        //あらかじめ決めておいた色コードを都道府県に合わせて取得(被らない)
        backgroundColor: prefColor[index],
        borderColor: prefColor[index],
      };
      setSelectedPrefectures(selectedCopy);
      setSeries([...series, resSeries]);
      setPrefecturesLoading(false);
      //checkboxが外されたらグラフ表示用のデータ配列からデータを削除
    } else {
      console.log(series.slice());
      const seriesCopy = series.slice();
      console.log(seriesCopy);
      for (let i = 0; i < seriesCopy.length; i++) {
        if (seriesCopy[i].label === prefectures[index].prefName) {
          seriesCopy.splice(i, 1);
        }
      }
      console.log(seriesCopy);

      setSelectedPrefectures(selectedCopy);
      setSeries(seriesCopy);
      setPrefecturesLoading(false);
    }
  };

  return (
    <div>
      <h1>日本総人口比較webサービス</h1>

      <Message />

      <h2>都道府県</h2>

      <div className="checkboxes">
        {Object.keys(prefectures).map((i) => {
          // 都道府県のチェックボックスを横に並べて表示させたい
          return (
            <PrefectureCheckbox
              key={prefectures[i].prefCode + 'checkbox_key'}
              checked={selectedPrefectures[prefectures[i].prefCode - 1]}
              disabled={prefecturesLoading}
              prefName={prefectures[i].prefName}
              prefCode={prefectures[i].prefCode}
              onChange={() => changeSelection(prefectures[i].prefCode - 1)}
            />
          );
        })}
      </div>

      <h2>総人口グラフ</h2>
      <div className="graph">
        <div className="graphInner">
          <Line
            data={{
              labels: [
                1960, 1965, 1970, 1975, 1980, 1985, 1990, 1995, 2000, 2005,
                2010, 2015, 2020, 2025, 2030, 2035, 2040, 2045,
              ],
              datasets: series,
            }}
            options={graphOption}
          />
        </div>
      </div>

      <div className="footer">created by kazuma iizumi</div>
    </div>
  );
};

export default App;
