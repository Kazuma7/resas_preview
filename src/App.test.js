import App from './App';
import React from 'react';
import renderer from 'react-test-renderer';

test('Screen display change notification', () => {
    const component = renderer.create(
        <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});