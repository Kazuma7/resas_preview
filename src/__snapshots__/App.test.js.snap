// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen display change notification 1`] = `
<div>
  <h1>
    日本総人口比較webサービス
  </h1>
  <div>
    <div
      className="main_message-box"
    >
      <p
        className="main_m"
      >
        💡ここはRESAS-APIを用いて都道府県ごとの総人口を調べ、比較することができるwebサービスです。
      </p>
    </div>
    <div
      className="sub_message_box"
    >
      <p
        className="sub_m"
      >
        RESAS-APIについてはこちら
      </p>
      <a
        className="sub_m_a1"
        href="https://opendata.resas-portal.go.jp/"
      >
        https://opendata.resas-portal.go.jp/
      </a>
      <p
        className="sub_m"
      >
        ソースコードはGitHubで公開中
      </p>
      <a
        className="sub_m_a"
        href="https://github.com/Kazuma7/resas-app"
      >
        https://github.com/Kazuma7/resas-app
      </a>
    </div>
  </div>
  <h2>
    都道府県
  </h2>
  <div
    className="checkboxes"
  />
  <h2>
    総人口グラフ
  </h2>
  <div
    className="graph"
  >
    <div
      className="graph-inner"
    >
      <canvas
        data-testid="canvas"
        height={150}
        onClick={[Function]}
        role="img"
        width={300}
      />
    </div>
  </div>
</div>
`;
